addons/hr_holidays/models/hr_holidays.py

```diff
@@ -0,0 +1,420 @@
+        ('date_check2', "CHECK ( \
+            (type='add') OR (date_from <= date_to) \
+            )", _("The start date must be anterior to the end date.")),
+        ('date_check', "CHECK ( number_of_days_temp >= 0 )",
+            _("The number of days must be greater than 0.")),
+    ]
+
+    @api.depends('number_of_days_temp')
+    def _compute_number_of_days(self):
+        for holiday in self:
+            if holiday.type == 'remove':
+                holiday.number_of_days = -holiday.number_of_days_temp
+            else:
+                holiday.number_of_days = holiday.number_of_days_temp
+
+    def _compute_can_reset(self):
+        """User can reset a leave request if it is its own leave request or if
+        he is an Hr Manager. """
+        if self.env.user.has_group('base.group_hr_manager'):
+            for holiday in self:
+                holiday.can_reset = True
+        else:
+            for holiday in self.filtered(lambda h: h.employee_id and h.employee_id.user_id == self.env.user):
+                holiday.can_reset = True
+
+    @api.constrains('date_from', 'date_to')
+    def _check_date(self):
+        for holiday in self:
+            domain = [
+                ('date_from', '<=', holiday.date_to),
+                ('date_to', '>=', holiday.date_from),
+                ('employee_id', '=', holiday.employee_id.id),
+                ('id', '!=', holiday.id),
+                ('state', 'not in', ['cancel', 'refuse']),
+            ]
+            leaves_taken = self.search_count(domain)
+            if leaves_taken:
+                raise ValidationError(_('You can not have 2 leaves that overlaps on same day!'))
+
+    @api.constrains('state', 'number_of_days_temp')
+    def _check_holiday(self):
+        for holiday in self.filtered(lambda h: h.holiday_type == 'employee' and h.type == 'remove' and h.employee_id and not h.holiday_status_id.limit):
+            leave_days = holiday.holiday_status_id.get_days(holiday.employee_id.id)[holiday.holiday_status_id.id]
+            if leave_days['remaining_leaves'] < 0 or leave_days['virtual_remaining_leaves'] < 0:
+                raise ValidationError(_('The number of remaining leaves is not sufficient for this leave type.\n Please verify also the leaves waiting for validation.'))
+
+    @api.multi
+    def name_get(self):
+        return [(leave.id, leave.name or _("%s on %s") % (leave.employee_id.name, leave.holiday_status_id.name)) for leave in self]
+
+    @api.onchange('holiday_type')
+    def _onchange_holiday_type(self):
+        if self.holiday_type == 'employee' and not self.employee_id:
+            self.employee_id = self.env['hr.employee'].search([('user_id', '=', self.env.uid)], limit=1).id
+        elif self.holiday_type != 'employee':
+            self.employee_id = False
+
+    @api.onchange('employee_id')
+    def _onchange_employee(self):
+        if self.employee_id:
+            self.department_id = self.employee_id.department_id.id
```

Comment:
No need to check **if self.employee_id** and better to assign recordset(s) instead of id(s).